### ISSUE TITLE
Can't decide between Date and Number cells

### ISSUE THREAD

**Issue by 2knu** [dhatim/fastexcel#110]
How can I decide if its a Date cell or just a number cell?
I want to read different excel sheets, where I don't know which one is a date cell. 

There is only a CellType Number and not a Date type. So I could try to read every number as Date, but than numbers will be read wrong for me, because they arent dates mostly.

So what I would expect is:

```
if (cell.getType() == CellType.Date)
{ cell.asDate(); }
else if(cell.getType() == CellType.Number)
{ cell.asNumber(); }
else 
{ cell.getText(); }
```

Would be nice when smt. like this will work :)

_Originally posted by @2knu in https://github.com/dhatim/fastexcel/issues/91#issuecomment-631576315_

**Comment by ochedru** [comment]
To achieve that, we would have to read the cell format (as replied in issue #91, date type does not exist) and apply complex checks (see [`DateUtil.isADateFormat()`](https://poi.apache.org/apidocs/4.1/org/apache/poi/ss/usermodel/DateUtil.html#isADateFormat-int-java.lang.String-)) to verify it is actually a date format.
This is a lot of work.

**Comment by sebastian-raubach** [comment]
I think this is an essential feature to know whether a cell actually contains a date or a number. I only realised this was an issue and am now seriously considering using a different library, because without this ability, I cannot use fastexcel anymore.

**Comment by lpyfm** [comment]
dates are stored internally in Excel as double values, so I don't think adding something like `CellType.Date` is possible.

With Apache POI, there's also no `Date` type, but there's the method `org.apache.poi.ss.usermodel.isCellDateFormatted`, which tells you if a cell is formatted as date. It would be great if fastexcel provides something similar.

**Comment by PGWelch** [comment]
> I think it is an essential feature to know whether a cell actually contains a date or a number. I only just realised this was an issue and am now seriously considering using a different library, because without this ability, I cannot use fastexcel anymore.

Agreed. I was going to use this library for reading excels as its more lightweight / less dependencies than POI, but given this issue I don't see how we could use it (which is a shame).

**Comment by perNyfelt** [comment]
This works fine now i think. If I define ReadingOptions as follows to get data format ids for each cell: 
`static final ReadingOptions OPTIONS = new ReadingOptions(true, true)`
and pass it to the ReadableWorkbook constructor:
`try (ReadableWorkbook workbook = new ReadableWorkbook(excelFile, OPTIONS))`

i can then check if the cell is date formatted by doing 
```groovy
// List of "magic" numbers corresponding to date formats in excel
static final List<Integer> DATE_FORMAT_IDS = List.of(14, 15, 16, 17, 18, 19, 20, 21, 22, 45, 46, 47, 165);
...
switch(cell.type) {
  case CellType.NUMBER -> {
    if (DATE_FORMAT_IDS.contains(cell.getDataFormatId()) {
      System.out.println("its a date:" + cell.asDate())
    } else {
      System.out.println("its a regular number:" + cell.asNumber())
    }
  }
  // handle other types...
}
```
The list of date format ids is probably not complete but "works for me" ;)
An alternative option would be to parse the dataFormat String but judging from the POI DateUtil that does this it is very complex and error prone.
